aibolit/clustering/clustering.py

```diff
@@ -0,0 +1,140 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Aibolit
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from collections import Counter
+from typing import Iterator, Tuple, List, Union
+from collections import OrderedDict
+from argparse import ArgumentParser
+from aibolit.utils.ast_builder import build_ast
+from aibolit.extract_method_baseline.extract_semantic import extract_method_statements_semantic  # type: ignore
+from aibolit.ast_framework import ASTNode, AST, ASTNodeType
+
+
+def check_is_common(dict_file, statement_1: Union[int, ASTNode], statement_2: Union[int, ASTNode]) -> bool:
+    joined_names: Counter = Counter(dict_file[statement_1] + dict_file[statement_2])
+    duplicates = {element: count for element, count in joined_names.items() if count > 1}.keys()
+    return len(list(duplicates)) >= 1
+
+
+def is_in_range(elem: int, values: List[int]) -> bool:
+    return elem >= values[0] and elem <= values[1]
+
+
+def process_statement(dict_file: OrderedDict, list_statements: List[Union[int, ASTNode]], step: int) -> List[List[int]]:
+    clusters: List[List[int]] = []
+    for stat_1 in list_statements:
+        stat_1_line = stat_1 if isinstance(stat_1, int) else stat_1.line
+        for stat_2 in list_statements[:stat_1_line + step]:
+            stat_2_line = stat_2 if isinstance(stat_2, int) else stat_2.line
+            if stat_1_line < stat_2_line and check_is_common(dict_file, stat_1, stat_2):
+                if len(clusters) != 0 and is_in_range(stat_1_line, clusters[-1]):
+                    if not is_in_range(stat_2_line, clusters[-1]):
+                        clusters[-1][1] = stat_2_line
+                else:
+                    clusters.append([stat_1_line, stat_2_line])
+    return clusters
+
+
+def SEMI_beta(dict_file: OrderedDict, method_len: int) -> dict:
```

Comment:
Could we specify return type? Dict of what?

Reply:
done